INT B-20011 B-20013 B-20014 HQ Role Queues, Searching, and Permissions

ajlusk opened the pull request:
## [B-20011](https://www13.v1host.com/USTRANSCOM38/assetdetail.v1?number=B-20011) & [B-20013](https://www13.v1host.com/USTRANSCOM38/assetdetail.v1?number=B-20013) & [B-20014](https://www13.v1host.com/USTRANSCOM38/assetdetail.v1?number=B-20014)

## Summary

Adds the following changes for the HQ role:
- Ability to create HQ user with HQ role (rather than TOO) from devlocal login page
- HQ User has it's own queues page containing all other office queues and searches
- HQ User has ability to make the necessary requests for all those queues
- HQ User can view all pages but not edit anything
- HQ User can leave customer service remarks
- Test for HQ User's queues page, mores tests to be added in later tickets

> [!NOTE]  
> In the Files Changed tab, the bulk of the additions will be collapsed due to being large diffs.

## Verification Steps for the Author

These are to be checked by the author.

- [ ] Tested in the Experimental environment (for changes to containers, app startup, or connection to data stores)
- [ ] Have the Agility acceptance criteria been met for this change?

## Verification Steps for Reviewers

These are to be checked by a reviewer.

- [ ] Has the branch been pulled in and checked out?
- [ ] Have the BL acceptance criteria been met for this change?
- [ ] Was the CircleCI build successful?
- [ ] Has the code been reviewed from a standards and best practices point of view?

### Setup to Run the Code

- [Instructions for starting storybook](https://transcom.github.io/mymove-docs/docs/frontend/setup/storybook)
- [Instructions for starting the MilMove application](https://transcom.github.io/mymove-docs/docs/getting-started/application-setup/)
- [Instructions for running tests](https://transcom.github.io/mymove-docs/docs/getting-started/development/testing)

### How to test

1. Ensure the HQ role FF is enabled
2. Access the office application
3. Create a new HQ user in the KKFA GBLOC (Switching GBLOCs will come in a later ticket)
4. Ensure the following functionality is available:
- [x] You are able to view, filter, sort, and open moves from the TOO's Queue
- [x] You are able to view, filter, sort, and open moves from the TIO's Queue
- [x] You are able to view, filter, sort, and open moves from the SC's Counseling Queue
- [x] You are able to view, filter, sort, and open moves from the SC's PPM Closeout Queue
- [x] You are able to search for moves
- [x] You are able to search for customers
5. Ensure when you open a move from any of the queues or the move search you are able to see everything but edit nothing except customer support remarks.
6. From the customer search you should not be able to open any of the users because those opened to edit forms that didn't have an existing way to be made read only.
7. Stop the server, disable the HQ Role feature flag in `.envrc` and `development.features.yaml`, run `direnv allow`, and restart your server.
8. You should now be met with the Invalid Permissions page when you log in as your previously created HQ user.
9. Try to navigate to various pages by manually updating the the path (`/moves-queue`, `/move/SITFUT/details`, etc.) and ensure you are shown the same invalid permissions page.

### Frontend

- [ ] There are no aXe warnings for UI.
- [ ] This works in [Supported Browsers and their phone views](https://transcom.github.io/mymove-docs/docs/adrs/Browser-Support/#minimum-browser-requirements) (Chrome, Firefox, Edge).
- [ ] There are no new console errors in the browser devtools.
- [ ] There are no new console errors in the test output.
- [ ] If this PR adds a new component to Storybook, it ensures the component is fully responsive, OR if it is intentionally not, a wrapping div using the `officeApp` class or custom `min-width` styling is used to hide any states the would not be visible to the user.
- [ ] This change meets the standards for [Section 508 compliance](https://www.ssa.gov/accessibility/andi/help/install.html).

### Backend

- [ ] Code follows the guidelines for [Logging](https://transcom.github.io/mymove-docs/docs/getting-started/development/logging).
- [ ] The requirements listed in [Querying the Database Safely](https://transcom.github.io/mymove-docs/docs/backend/guides/golang-guide#querying-the-database-safely) have been satisfied.

### Database

#### Any new migrations/schema changes:

- [ ] Follows our guidelines for [Zero-Downtime Deploys](https://transcom.github.io/mymove-docs/docs/backend/setup/database-migrations#zero-downtime-migrations).
- [ ] Have been communicated to #g-database.
- [ ] Secure migrations have been tested following the instructions in our [docs](https://transcom.github.io/mymove-docs/docs/backend/setup/database-migrations#secure-migrations).

## Screenshots

robot-mymove commented:
<!--
  0 failure: 
  1 warning:  Files located in ...
  
  
  DangerID: danger-id-Danger;
-->


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  Files located in legacy directories (src/shared or src/scenes) have
been edited. Are you sure you don’t want to also relocate them to the new [file structure](https://transcom.github.io/mymove-docs/docs/dev/contributing/frontend/frontend#file-layout--naming)?

View the [frontend file org ADR](https://github.com/transcom/mymove/blob/main/docs/adr/0048-frontend-file-org.md) for more information
  </td>
    </tr>
  </tbody>
</table>



<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 4d376623c9d505c746e9ac1d40e1908a4e39e8dc
</p>

antgmann commented on pkg/services/office_user/customer/customer_searcher.go:
```diff
@@ -49,33 +53,31 @@ func (s customerSearcher) SearchCustomers(appCtx appcontext.AppContext, params *
 	var query *pop.Query
 
-	if appCtx.Session().Roles.HasRole(roles.RoleTypeServicesCounselor) {
-		rawquery := `SELECT DISTINCT ON (id)
```
Is this if removed because the query will already not be called due to line 40-43 above? Or should it check for service counselor || HQ role?

ajlusk replied:
When I first tried to make the request an an HQ user it was failing ungracefully (null pointer error resulting in a 500). I originally just added || HQ but that error wasn't sitting right with me. So instead I added lines 40-43 and lines 92-97 in `pkg/handlers/ghcapi/customer.go`. That made the condition I removed redundant. This allows the HQ user to make the request and makes the request return a 403 without a backend null pointer error when roles other than SC or HQ make the request.

r-mettler commented:
<img width="697" alt="image" src="https://github.com/transcom/mymove/assets/149000222/1871c8ea-e53a-4fd0-8374-20df6b2d48b9">
Can add edit and delete customer remarks as intended.
All other features disabled as intended.

r-mettler commented:
Looks like we need to disable access for when FF for HQ role is set to false. I was still able to access the HQ queue when I switched the flag to false. I ran direnv allow and restarted the server and client and verified the flag switched because I can't create a new Office account with HQ role in the Admin app. 😢

ajlusk commented:
> Looks like we need to disable access for when FF for HQ role is set to false. I was still able to access the HQ queue when I switched the flag to false. I ran direnv allow and restarted the server and client and verified the flag switched because I can't create a new Office account with HQ role in the Admin app. We can ask Danny what he thinks. With the FF off they can't create new HQ role accounts but if they already have an existing account they will be able to login still. I don't know how much that matters since everything is read only but comments so its not like they can do much damage at this point.

Per this comment and PO agreement I have updated this PR so an existing HQ user will be shown an Invalid Permissions page when they log in and when they visit any other path. The testing instructions have been updated to reflect that.

antgmann commented:
> > Looks like we need to disable access for when FF for HQ role is set to false. I was still able to access the HQ queue when I switched the flag to false. I ran direnv allow and restarted the server and client and verified the flag switched because I can't create a new Office account with HQ role in the Admin app. We can ask Danny what he thinks. With the FF off they can't create new HQ role accounts but if they already have an existing account they will be able to login still. I don't know how much that matters since everything is read only but comments so its not like they can do much damage at this point.
> 
> Per this comment and PO agreement I have updated this PR so an existing HQ user will be shown an Invalid Permissions page when they log in and when they visit any other path. The testing instructions have been updated to reflect that.

Tested and verified!

antgmann reviewed (approved):
Functionally tested and reviewed, looks great! FF prevents users from accessing pages when turned back off. Cannot edit on any pages, read-only.

r-mettler reviewed (approved):
Changes look good. Tested latest update and can't login with a HQ user if the FF is turned off.
